plugins/janus_videoroom.c

```diff
@@ -1585,10 +1587,19 @@ static int janus_videoroom_access_room(json_t *root, gboolean check_secret, gboo
 			return error_code;
 		}
 	}
-	if(check_pin) {
+	if(check_join) {
 		char error_cause2[100];
-		JANUS_CHECK_SECRET((*videoroom)->room_pin, root, "pin", error_code, error_cause2,
-			JANUS_VIDEOROOM_ERROR_MISSING_ELEMENT, JANUS_VIDEOROOM_ERROR_INVALID_ELEMENT, JANUS_VIDEOROOM_ERROR_UNAUTHORIZED);
+		/* signed tokens bypass pin validation */
+		json_t *token = json_object_get(root, "token");
+		if(token && janus_auth_is_stored_mode()) {
```

Comment:
This doesn't look right. This kind of exchanges should go through the core-plugin API. Violating the API will cause trouble.

Reply:
Besides, I'm not comfortable in getting rid of the pin mechanism altogether.

Reply:
core-plugin API here means something like "defined in utils.h"?
You mean a macro similar to JANUS_CHECK_SECRET?

Reply:
in general the videoroom implementation I am writing right now is a pretty wonky sketch; the overlaps between `secret`, `pin`, `allowed` and `token` need to be cleared more. Any suggestions or help (exactly like this comment) are very welcome.

By the way, the only check for `allowed` tokens is in the publisher path for join if i see that correctly, is that right? does this mean people can subscribe to rooms without pin?

Reply:
No, as in defined in plugin.h: https://github.com/meetecho/janus-gateway/blob/master/plugins/plugin.h
Think of what methods like `push_event` or `events_is_enabled` do.

The tokens are something the core stores, and side access to that should not happen outside of what a specific API to interact with the core provides and allows. Besides, I'm not even sure that with the dynamic loading of libraries accessing via auth.h is guaranteed to work.

Reply:
ah, gotcha. I skimmed `plugin.h` but didn't realise there was a part for calling into the other direction. The code as it is right now works, but I suppose there might be problems if the library and janus aren't built together. I will expose the token API there then.

>The tokens are something the core stores, and side access to that should not happen outside of what a specific API to interact with the core provides and allows.

I am not sure I understand this part. With the signature based approach no tokens are stored - the `janus_auth_check_signature*` API is exposed only for convenience and uniformity of tokens, it could be implemented separately by plugins too for all they care.  
The tokens going into the check also come from a source the plugin controls, not out of the general janus transport. To illustrate, here is what my `joinandconfigure` payload looks like at the moment:
```json
{
  "janus": "message",
  "body": {
    "request": "joinandconfigure",
    "ptype": "publisher",
    "room": 5273654000077434,
    "id": 610,
    "display": "asd",
    "audio": true,
    "video": true,
    "data": true,
    "bitrate": 32,
    "token": "1519907724,room=52736540000774342:guu1NRpABI8L492R0Wb+0CaFY2Y="
  },
  "transaction": "ysX4m3zaoMNY",
  "token": "1519907724,janus.plugin.videoroom:QW2EdMHiQo1Hk5FU0fybXxPacB8=",
  "session_id": 3355679644666351,
  "handle_id": 5009926784814890
}
```

there are two separate tokens generated by my API endpoint, I call them `token` and `roomToken`:
```js
        res.send(JSON.stringify({
          turnCredentials: getTURNCredentials(user._id),
          rtcCredentials: {
            token: getJanusToken(['janus.plugin.videoroom']),
            room,
            roomToken: getJanusToken([`room=${room}`]),
            // pin,
          },
        }));
```

Reply:
>Besides, I'm not comfortable in getting rid of the pin mechanism altogether.

Do you mean you want to preserve backwards compatibility for pin-based authentication or that you dislike the way in which they interact in this draft?
With the current code pin works exactly as before unless `token_auth_secret` is provided *and* a **valid** token is presented by the client.

As mentioned above I'm also not very happy with this current implementation, but I am not sure which is the way to go. I was thinking of using `check_token`, but in the end it's pretty pointless to require the application server to enable/disable token authentication on a per-room basis because if the server wants restrict token access it should just not sign tokens for those rooms.